### DISCUSSION TITLE
lsolve, usolve for non-square matrices?

### DISCUSSION THREAD

**Discussion by mike239x** [josdejong/mathjs#2754] [Ideas]
It seems [`lsolve`](http://mathjs.org/docs/reference/functions/lsolve.html) and [`usolve`](mathjs.org/docs/reference/functions/usolve.html)  are only working for the square matrices.

Why is it so? I think it should work for non-square matrices as well.

Well, the important one is the `usolve`, because if you use LU decomposition 
(in my case using [`lup`](http://mathjs.org/docs/reference/functions/lup.html)), you get square L matrix and non-square U matrix.

**Comment by mike239x** [comment]
This is pretty easy to fix by simple adding some zeros to make the matrix square either adding/removing zeros here and there.

**Comment by josdejong** [comment]
I don't know why the current implementation only supports square matrices, but if it's easy to add support for non-square matrices that would be a great extension.

@mike239x would you be interested in giving it a try and extend `lsolve` and `usolve`?

**Comment by mike239x** [comment]
@josdejong Right now I created my self-made implementation with some additional restrictions, so that it would be enough for me, later I can do that.  Right now kinda busy with my bachelor thesis, my bad.
You can assign this task to me if you want.

**Comment by josdejong** [comment]
:+1: 

Let's just keep it open. If anyone can pick this up before Mike can please drop a message here.

**Comment by a8t** [comment]
I wouldn't mind attempting, but I encourage others to hop on as well. Been a few years since I graduated, haven't done much matrix math since... 😅

**Comment by cshaa** [comment]
This should be relatively easy to implement, the algorithms needed are already implemented in [`lsolveAll`](https://github.com/josdejong/mathjs/blob/develop/src/function/algebra/solver/lsolveAll.js) and [`usolveAll`](https://github.com/josdejong/mathjs/blob/develop/src/function/algebra/solver/usolveAll.js). They generalize well to the case of rectangular matrices, as described [here](https://www.overleaf.com/read/csvgqdxggyjv), somebody just needs to implement it :)

If anyone's up for the task, I'll happily help them understand the current code and what needs to be done exactly.
